## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#759e80a0-9064-41e3-8d87-7c9646b9dddb) |
| mint-gateway-azure.sh | :x: [more...](#a57b99b9-09be-495d-a0de-be1f4f7ef188) |
| mint-gateway-nas.sh | :x: [more...](#4528d423-8f77-4ae3-bd5f-7deff6ff7ba0) |
<a name="4528d423-8f77-4ae3-bd5f-7deff6ff7ba0"></a>
### 9426-99f1ac8/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31175
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 89effd2ae730:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 1 seconds
(5/15) Running awscli tests ... done in 1 minutes and 5 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 22 seconds
(8/15) Running minio-dotnet tests ... done in 32 seconds
(9/15) Running minio-go tests ... FAILED in 26 seconds
{}
(9/15) Running minio-java tests ... FAILED in 38 seconds
{
  "name": "minio-java",
  "function": "enableObjectLegalHold(String bucketName, String objectName, String versionId)",
  "duration": 26,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidRequest, message = Bucket is missing ObjectLockConfiguration, bucketName = minio-java-test-3fu1k79, objectName = minio-java-test-4u4q9c, resource = /minio-java-test-3fu1k79/minio-java-test-4u4q9c, requestId = 16085839D0455136, hostId = 34f51f9a-a90e-43ef-a31d-026acff31b21)\nrequest={method=PUT, url=http://minio-dev6.minio.io:31175/minio-java-test-3fu1k79/minio-java-test-4u4q9c?legal-hold=, headers=Host: minio-dev6.minio.io:31175\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: erOisiAmIIzsh/Ze2QqGXw==\nx-amz-content-sha256: 96b73c95a8d33e664ab2170e095025b47ebd55978bb71cebd6a51e394bf96722\nx-amz-date: 20200423T042433Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200423/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 382\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 16085839D0455136\nX-Xss-Protection: 1; mode=block\nDate: Thu, 23 Apr 2020 04:24:33 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.executePut(MinioClient.java:1424), io.minio.MinioClient.enableObjectLegalHold(MinioClient.java:3568), FunctionalTest.enableObjectLegalHold_test(FunctionalTest.java:3303), FunctionalTest.runTests(FunctionalTest.java:4079), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 31 seconds
(10/15) Running minio-py tests ... done in 1 minutes and 4 seconds
(11/15) Running s3cmd tests ... done in 17 seconds
(12/15) Running s3select tests ... done in 1 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="a57b99b9-09be-495d-a0de-be1f4f7ef188"></a>
### 9426-99f1ac8/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31176
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bbd12637e259:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 10 seconds
(4/15) Running aws-sdk-ruby tests ... done in 14 seconds
(5/15) Running awscli tests ... FAILED in 51 seconds
{
  "name": "awscli",
  "duration": 3759,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:31176 s3api copy-object --bucket awscli-mint-test-bucket-17164 --storage-class REDUCED_REDUNDANCY --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-17164/datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (NotImplemented) when calling the CopyObject operation: A header you provided implies functionality that is not implemented"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 6 minutes and 17 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 27 seconds
(8/15) Running minio-go tests ... FAILED in 9 minutes and 48 seconds
{}
(8/15) Running minio-java tests ... FAILED in 26 minutes and 15 seconds
{
  "name": "minio-java",
  "function": "enableObjectLegalHold(String bucketName, String objectName, String versionId)",
  "duration": 1701,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidRequest, message = Bucket is missing ObjectLockConfiguration, bucketName = minio-java-test-18ueaah, objectName = minio-java-test-p7k3fv, resource = /minio-java-test-18ueaah/minio-java-test-p7k3fv, requestId = 16085A89D8307A7D, hostId = 583888d5-c5a4-40d4-b977-cfadc966b226)\nrequest={method=PUT, url=http://minio-dev6.minio.io:31176/minio-java-test-18ueaah/minio-java-test-p7k3fv?legal-hold=, headers=Host: minio-dev6.minio.io:31176\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: erOisiAmIIzsh/Ze2QqGXw==\nx-amz-content-sha256: 96b73c95a8d33e664ab2170e095025b47ebd55978bb71cebd6a51e394bf96722\nx-amz-date: 20200423T050656Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200423/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 382\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 16085A89D8307A7D\nX-Xss-Protection: 1; mode=block\nDate: Thu, 23 Apr 2020 05:06:56 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.executePut(MinioClient.java:1424), io.minio.MinioClient.enableObjectLegalHold(MinioClient.java:3568), FunctionalTest.enableObjectLegalHold_test(FunctionalTest.java:3303), FunctionalTest.runTests(FunctionalTest.java:4079), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(8/15) Running minio-js tests ... FAILED in 3 minutes and 40 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-9dd72646-c928-447e-bb44-12f697fdd701, recursive:true",
  "duration": 168,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(8/15) Running minio-py tests ... done in 9 minutes and 51 seconds
(9/15) Running s3cmd tests ... FAILED in 1 minutes and 50 seconds
{
  "name": "s3cmd",
  "duration": "2343",
  "function": "test_copy_object",
  "status": "FAIL",
  "error": ""
}
(9/15) Running s3select tests ... done in 22 seconds
(10/15) Running security tests ... done in 0 seconds

Executed 10 out of 15 tests successfully.
```
<a name="759e80a0-9064-41e3-8d87-7c9646b9dddb"></a>
### 9426-99f1ac8/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30902
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 38e4c30dbdc3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-f5v0hsnad5gi4i",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365405,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16085859FB6A5AAA, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-f5v0hsnad5gi4i",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365405,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16085859FB6A5AAA, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-f5v0hsnad5gi4i",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365405,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 16085859FB6A5AAA, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-ruby tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-ruby",
  "function": "listBuckets",
  "args": {},
  "duration": 2136.03,
  "error": "Server not initialized, please try again.",
  "status": "FAIL"
}
(1/15) Running awscli tests ... FAILED in 11 seconds
{
  "name": "awscli",
  "duration": 9536,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}
(1/15) Running healthcheck tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": null,
  "duration": 70,
  "function": "testReadinessEndpoint",
  "message": "GET /minio/health/ready returned non OK status",
  "name": "healthcheck",
  "status": "FAIL"
}
(1/15) Running mc tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": null,
  "duration": 70,
  "function": "testReadinessEndpoint",
  "message": "GET /minio/health/ready returned non OK status",
  "name": "healthcheck",
  "status": "FAIL"
}
(1/15) Running minio-dotnet tests ... done in 3 seconds
(2/15) Running minio-go tests ... FAILED in 1 minutes and 26 seconds
{
  "args": {
    "bucketName": "minio-go-test-6fxckbneod22uw11",
    "region": "eu-west-1"
  },
  "duration": 85795,
  "error": "Server not initialized, please try again.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(2/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(String bucketName)",
  "duration": 235,
  "status": "FAIL",
  "error": "java.lang.IllegalArgumentException: unknown error code string 'XMinioServerNotInitialized' >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.execute(MinioClient.java:1134), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.makeBucket_test1(FunctionalTest.java:247), FunctionalTest.runTests(FunctionalTest.java:4005), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(2/15) Running minio-js tests ... FAILED in 1 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 4,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(2/15) Running minio-py tests ... FAILED in 10 minutes and 1 seconds
{
  "name": "minio-py:test_make_bucket_default_region",
  "function": "make_bucket(bucket_name, location)",
  "args": {
    "bucket_name": "minio-py-test-e7f4f318-f744-4fb9-af60-2380dfad1a6e",
    "location": "default value ('us-east-1')"
  },
  "duration": 600507,
  "message": "HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /minio-py-test-e7f4f318-f744-4fb9-af60-2380dfad1a6e/ (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 188, in test_make_bucket_default_region\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 371, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /minio-py-test-e7f4f318-f744-4fb9-af60-2380dfad1a6e/ (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1968, in main\n    test_make_bucket_default_region(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 196, in test_make_bucket_default_region\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /minio-py-test-e7f4f318-f744-4fb9-af60-2380dfad1a6e/ (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(2/15) Running s3cmd tests ... FAILED in 45 seconds
{
  "name": "s3cmd",
  "duration": "45424",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(2/15) Running s3select tests ... FAILED in 20 minutes and 1 seconds
{
  "name": "minio-py:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-309c6504-c417-4069-92ce-4502ecde1e10"
  },
  "duration": 1200766,
  "message": "HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /s3select-test-309c6504-c417-4069-92ce-4502ecde1e10?location= (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 144, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 371, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /s3select-test-309c6504-c417-4069-92ce-4502ecde1e10/ (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 183, in test_csv_input_custom_quote_char\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /s3select-test-309c6504-c417-4069-92ce-4502ecde1e10/ (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 186, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 446, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1984, in _url_open\n    region = self._get_bucket_region(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1919, in _get_bucket_region\n    region = self._get_bucket_location(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1950, in _get_bucket_location\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /s3select-test-309c6504-c417-4069-92ce-4502ecde1e10?location= (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 310, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"./tests.py\", line 188, in test_csv_input_custom_quote_char\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev7.minio.io', port=30902): Max retries exceeded with url: /s3select-test-309c6504-c417-4069-92ce-4502ecde1e10?location= (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(2/15) Running security tests ... done in 0 seconds

Executed 2 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/disk-cache-backend.go
+++ b/cmd/disk-cache-backend.go
@@ -21,10 +21,10 @@ import (
 	"context"
 	"crypto/rand"
 	"encoding/hex"
+	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
-	"log"
 	"net/http"
 	"os"
 	"strings"
@@ -228,6 +228,10 @@ func (c *diskCache) toClear() uint64 {
 	return bytesToClear(int64(di.Total), int64(di.Free), uint64(c.quotaPct), uint64(c.lowWatermark))
 }
 
+var (
+	errDoneForNow = errors.New("done for now")
+)
+
 // Purge cache entries that were not accessed.
 func (c *diskCache) purge(ctx context.Context) {
 	if c.diskUsageLow() {
@@ -263,25 +267,24 @@ func (c *diskCache) purge(ctx context.Context) {
 		}
 		return fm
 	}
-	objDirs, err := ioutil.ReadDir(c.dir)
-	if err != nil {
-		log.Fatal(err)
-	}
 
-	for _, obj := range objDirs {
-		if obj.Name() == minioMetaBucket {
-			continue
+	filterFn := func(name string, typ os.FileMode) error {
+		if name == minioMetaBucket {
+			// Proceed to next file.
+			return nil
 		}
 
-		cacheDir := pathJoin(c.dir, obj.Name())
+		cacheDir := pathJoin(c.dir, name)
 		meta, _, numHits, err := c.statCachedMeta(ctx, cacheDir)
 		if err != nil {
 			// delete any partially filled cache entry left behind.
 			removeAll(cacheDir)
-			continue
+			// Proceed to next file.
+			return nil
 		}
+
 		// stat all cached file ranges and cacheDataFile.
-		cachedFiles := fiStatFn(meta.Ranges, cacheDataFile, pathJoin(c.dir, obj.Name()))
+		cachedFiles := fiStatFn(meta.Ranges, cacheDataFile, pathJoin(c.dir, name))
 		objInfo := meta.ToObjectInfo("", "")
 		cc := cacheControlOpts(objInfo)
 		for fname, fi := range cachedFiles {
@@ -291,9 +294,11 @@ func (c *diskCache) purge(ctx context.Context) {
 						logger.LogIf(ctx, err)
 					}
 					scorer.adjustSaveBytes(-fi.Size())
+
 					// break early if sufficient disk space reclaimed.
 					if c.diskUsageLow() {
-						return
+						// if we found disk usage is already low, we return nil filtering is complete.
+						return errDoneForNow
 					}
 				}
 				continue
@@ -305,12 +310,24 @@ func (c *diskCache) purge(ctx context.Context) {
 		if err != nil || (fi.ModTime().Before(expiry) && len(cachedFiles) == 0) {
 			removeAll(cacheDir)
 			scorer.adjustSaveBytes(-fi.Size())
-			continue
+			// Proceed to next file.
+			return nil
 		}
+
+		// if we found disk usage is already low, we return nil filtering is complete.
 		if c.diskUsageLow() {
-			return
+			return errDoneForNow
 		}
+
+		// Proceed to next file.
+		return nil
 	}
+
+	if err := readDirFilterFn(c.dir, filterFn); err != nil {
+		logger.LogIf(ctx, err)
+		return
+	}
+
 	for _, path := range scorer.fileNames() {
 		removeAll(path)
 		slashIdx := strings.LastIndex(path, SlashSeparator)

--- a/cmd/posix-list-dir_other.go
+++ b/cmd/posix-list-dir_other.go
@@ -1,7 +1,7 @@
 // +build plan9 solaris
 
 /*
- * MinIO Cloud Storage, (C) 2016, 2017, 2018 MinIO, Inc.
+ * MinIO Cloud Storage, (C) 2016-2020 MinIO, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -30,6 +30,44 @@ import (
 // Return all the entries at the directory dirPath.
 func readDir(dirPath string) (entries []string, err error) {
 	return readDirN(dirPath, -1)
+}
+
+// readDir applies the filter function on each entries at dirPath, doesn't recurse into
+// the directory itself.
+func readDirFilterFn(dirPath string, filter func(name string, typ os.FileMode) error) error {
+	d, err := os.Open(dirPath)
+	if err != nil {
+		// File is really not found.
+		if os.IsNotExist(err) {
+			return errFileNotFound
+		}
+
+		// File path cannot be verified since one of the parents is a file.
+		if strings.Contains(err.Error(), "not a directory") {
+			return errFileNotFound
+		}
+		return err
+	}
+	defer d.Close()
+
+	maxEntries := 1000
+	for {
+		// Read up to max number of entries.
+		fis, err := d.Readdir(maxEntries)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		for _, fi := range fis {
+			if err = filter(fi.Name(), fi.Mode()); err == errDoneForNow {
+				// filtering requested to return by caller.
+				return nil
+			}
+		}
+	}
+	return nil
 }
 
 // Return N entries at the directory dirPath. If count is -1, return all entries

--- a/cmd/posix-list-dir_unix.go
+++ b/cmd/posix-list-dir_unix.go
@@ -1,7 +1,7 @@
 // +build linux,!appengine darwin freebsd netbsd openbsd
 
 /*
- * MinIO Cloud Storage, (C) 2016, 2017, 2018 MinIO, Inc.
+ * MinIO Cloud Storage, (C) 2016-2020 MinIO, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -73,6 +73,59 @@ func parseDirEnt(buf []byte) (consumed int, name string, typ os.FileMode, err er
 // Return all the entries at the directory dirPath.
 func readDir(dirPath string) (entries []string, err error) {
 	return readDirN(dirPath, -1)
+}
+
+// readDir applies the filter function on each entries at dirPath, doesn't recurse into
+// the directory itself.
+func readDirFilterFn(dirPath string, filter func(name string, typ os.FileMode) error) error {
+	fd, err := syscall.Open(dirPath, 0, 0)
+	if err != nil {
+		if os.IsNotExist(err) || isSysErrNotDir(err) {
+			return errFileNotFound
+		}
+		if os.IsPermission(err) {
+			return errFileAccessDenied
+		}
+		return err
+	}
+	defer syscall.Close(fd)
+
+	buf := make([]byte, blockSize) // stack-allocated; doesn't escape
+	boff := 0                      // starting read position in buf
+	nbuf := 0                      // end valid data in buf
+
+	for {
+		if boff >= nbuf {
+			boff = 0
+			nbuf, err = syscall.ReadDirent(fd, buf)
+			if err != nil {
+				if isSysErrNotDir(err) {
+					return errFileNotFound
+				}
+				return err
+			}
+			if nbuf <= 0 {
+				break
+			}
+		}
+		consumed, name, typ, err := parseDirEnt(buf[boff:nbuf])
+		if err != nil {
+			return err
+		}
+		boff += consumed
+		if name == "" || name == "." || name == ".." {
+			continue
+		}
+		if typ&os.ModeSymlink == os.ModeSymlink {
+			continue
+		}
+		if err = filter(name, typ); err == errDoneForNow {
+			// filtering requested to return by caller.
+			return nil
+		}
+	}
+
+	return err
 }
 
 // Return count entries at the directory dirPath and all entries

--- a/cmd/posix-list-dir_windows.go
+++ b/cmd/posix-list-dir_windows.go
@@ -1,7 +1,7 @@
 // +build windows
 
 /*
- * MinIO Cloud Storage, (C) 2016, 2017, 2018 MinIO, Inc.
+ * MinIO Cloud Storage, (C) 2016-2020 MinIO, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -27,6 +27,69 @@ import (
 // Return all the entries at the directory dirPath.
 func readDir(dirPath string) (entries []string, err error) {
 	return readDirN(dirPath, -1)
+}
+
+// readDir applies the filter function on each entries at dirPath, doesn't recurse into
+// the directory itself.
+func readDirFilterFn(dirPath string, filter func(name string, typ os.FileMode) error) error {
+	d, err := os.Open(dirPath)
+	if err != nil {
+		// File is really not found.
+		if os.IsNotExist(err) {
+			return errFileNotFound
+		}
+
+		// File path cannot be verified since one of the parents is a file.
+		if strings.Contains(err.Error(), "not a directory") {
+			return errFileNotFound
+		}
+		return err
+	}
+	defer d.Close()
+
+	st, err := d.Stat()
+	if err != nil {
+		return err
+	}
+	// Not a directory return error.
+	if !st.IsDir() {
+		return errFileAccessDenied
+	}
+
+	data := &syscall.Win32finddata{}
+
+	for {
+		e := syscall.FindNextFile(syscall.Handle(d.Fd()), data)
+		if e != nil {
+			if e == syscall.ERROR_NO_MORE_FILES {
+				break
+			} else {
+				err = &os.PathError{
+					Op:   "FindNextFile",
+					Path: dirPath,
+					Err:  e,
+				}
+				return err
+			}
+		}
+		name := syscall.UTF16ToString(data.FileName[0:])
+		if name == "" || name == "." || name == ".." { // Useless names
+			continue
+		}
+		if data.FileAttributes&syscall.FILE_ATTRIBUTE_REPARSE_POINT != 0 {
+			continue
+		}
+		var typ os.FileMode = 0 // regular file
+		if data.FileAttributes&syscall.FILE_ATTRIBUTE_DIRECTORY != 0 {
+			typ = os.ModeDir
+		}
+		if err = filter(name, typ); err == errDoneForNow {
+			// filtering requested to return by caller.
+			return nil
+		}
+	}
+
+	return err
 }
 
 // Return N entries at the directory dirPath. If count is -1, return all entries

